compiler/rustc_codegen_cranelift/src/base.rs

```diff
@@ -780,6 +780,9 @@ fn codegen_stmt<'tcx>(
                             let variant_dest = lval.downcast_variant(fx, variant_index);
                             (variant_index, variant_dest, active_field_index)
                         }
+                        mir::AggregateKind::Coroutine(_, _) => {
+                            (FIRST_VARIANT, lval.downcast_variant(fx, FIRST_VARIANT), None)
```

Comment:
The fact that the unresumed state is the first variant, that's just an assumption that this code is making, right?

I'm not certain how heavily that assumption is baked into all the rest of the code. I do see that the variant `CoroutineArgs::UNRESUMED` is assigned the value 0, but I don't know if that value is set in stone.

(Maybe all I'm looking for is an assertion ensuring `FIRST_VARIANT` *is* the variant for Unresumed on all these places where you downcast to FIRST_VARIANT.)

Reply:
That's correct. `UNRESUMED` is used across, except in the codegen backends which is `FIRST_VARIANT`. It appears that the first variant is assumed for coroutines and `struct`s, though this fact was not obvious at the first glance.